### PR TITLE
warn about uninitialized multi-variant enums (invalid_value lint)

### LINKED ISSUE #73608: "incorrect use of uninit memory" lint should fire for multi-variant enum
I tried this code:

```rust
#[derive(Copy, Clone, Debug)]
enum Fruit {
    Apple,
    _Banana,
}

fn foo() -> Fruit {
    unsafe {
        let mut r = mem::uninitialized();
        ptr::write(&mut r as *mut Fruit, Fruit::Apple);
        r
    }
}
```

I expected to see this happen: It should warn about `Fruit` not working with `mem::uninitialized`.

Instead, this happened: There is no diagnostic.

The problem here is that not *all* enums should cause the warning. If the enum only has one variant, being uninitialized could actually be fine (but variants that are uninhabited *and* ZST do not count). This is easy to check at run-time where we have the full layout (so we panic), but statically we just have the type and that makes it harder.

As a start, we could count the fieldless variants; if there are at least 2 then the enum definitely needs a tag and thus may not remain uninitialized.

Cc @stepancheg

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/73608

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1922,6 +1922,14 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
             None
         }
 
+        /// Test if this enaum has several actually "existing" variants.
```
Comment (lcnr) [inline src/librustc_lint/builtin.rs:1925]:
```suggestion
        /// Test if this enum has several actually "existing" variants.
```

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -2001,10 +2009,17 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
...
-                        // uninhabited, we might actually do layout like for a single-variant
-                        // enum, and then even leaving them uninitialized could be okay.
-                        _ => None, // Conservative fallback for multi-variant enum.
+                        // Multi-variant enum.
+                        _ => {
+                            if init == InitKind::Uninit && is_multi_variant(adt_def) {
+                                let span = tcx.def_span(adt_def.did);
+                                Some(("enums have to be initialized to a variant".to_string(), Some(span)))
+                            } else {
+                                // In principle, for zero-initialization we could figure out which variant correpsonds
```
Comment (lcnr) [inline src/librustc_lint/builtin.rs:2018]:
```suggestion
                                // In principle, for zero-initialization we could figure out which variant corresponds
```

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -2001,10 +2009,17 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
...
+                                // to tag 0, and check that... but for now we just accept all zero-initialization.
```
Comment (lcnr) [inline src/librustc_lint/builtin.rs:2019]:
```suggestion
                                // to tag 0, and check that... but for now we just accept all zero-initializations.
```

**Review by lcnr — APPROVED** [review]

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after other comments resolved.

**Comment by RalfJung** [comment]
*oops* so many typos...

@bors r=davidtwco

**Comment by davidtwco** [comment]
@RalfJung if it isn't too much trouble, could you squash those last two commits?

**Comment by RalfJung** [comment]
@bors r-
@davidtwco  Sure. do you mean squash them into one (so there are 2 commits overall), or squash them both into the first (so there is one commit overall)?

**Comment by davidtwco** [comment]
> @bors r-
> @davidtwco Sure. do you mean squash them into one (so there are 2 commits overall), or squash them both into the first (so there is one commit overall)?

Squash both into the first commit.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**